strapi/strapi
# enhancement: U&P config register.allowedFields defaults to empty array

### What does it do?

- When plugins.js `users-permissions.register.allowedFields` value is not set, it is treated as an empty array, meaning that no fields other than the required strapi fields are accepted on user registration (/auth/local/register). Previously, an empty value meant that all user-created fields would be accepted and only block internal Strapi fields from being accepted.
- invalid fields now throws a ValidationError (400) instead of being silently removed
- removes the warning on startup that is no longer needed.

### Why is it needed?

There was a security issue in Strapi v4 that required the addition of the allowedFields config option, but to prevent a breaking change it required user action to ensure any custom u&p user fields were protected.

In v5, the allowedFields option will be required in order to accept any custom u&p user fields.

### How to test it?

Unit tests of the register method have been provided to test normal operation, error condition, and allowedFields override

To test manually:

- add an additional field to your u&p User schema
- submit a new user registration to /auth/local/register including that field
- the project should NOT have a config.js users-permissions.register.allowedFields value set

Without this PR:
- when starting the strapi server, a warning will be logged that that field is not protected
- the additional field will be accepted and inserted into the new user record
- if you create an empty array for users-permissions.register.allowedFields, the user will still be created successfully (200 status) but the field will not be accepted

With this PR:
- the additional field will not be accepted; instead a 400 error should be received

### Related issue(s)/PR(s)

The original feature: https://github.com/strapi/strapi/pull/17804
